callstats-java-sdk/src/main/java/io/callstats/sdk/CallStatsHttpClient.java

```diff
@@ -0,0 +1,226 @@
+import java.net.URISyntaxException;
+
+import org.apache.http.HttpResponse;
+import org.apache.http.client.ClientProtocolException;
+import org.apache.http.client.HttpClient;
+import org.apache.http.client.methods.HttpPost;
+import org.apache.http.client.methods.HttpUriRequest;
+import org.apache.http.concurrent.FutureCallback;
+import org.apache.http.entity.StringEntity;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.impl.conn.PoolingHttpClientConnectionManager;
+import org.apache.http.impl.nio.client.CloseableHttpAsyncClient;
+import org.apache.http.impl.nio.client.HttpAsyncClients;
+import org.apache.http.impl.nio.conn.PoolingNHttpClientConnectionManager;
+import org.apache.http.impl.nio.reactor.DefaultConnectingIOReactor;
+import org.apache.http.impl.nio.reactor.IOReactorConfig;
+import org.apache.http.message.BasicHeader;
+import org.apache.http.nio.client.HttpAsyncClient;
+import org.apache.http.nio.reactor.ConnectingIOReactor;
+import org.apache.http.nio.reactor.IOReactorException;
+
+/**
+*
+* @author Karthik Budigere
+*/
+public class CallStatsHttpClient {
+	
+	private static String BASE_URL = "http://127.0.0.1:3000";
+	
+	private String appId;
+	private String authToken;
+	private int numRetries = 3;
+	private HttpClient httpclient;
+	private HttpAsyncClient httpAsyncClient;
+	
+	public HttpClient getHttpclient() {
+		return httpclient;
+	}
+	
+	public void setHttpclient(HttpClient httpclient) {
+		this.httpclient = httpclient;
+	}
+
+	public int getNumRetries() {
+		return numRetries;
+	}
+	
+	public void setNumRetries(int numRetries) {
+		this.numRetries = numRetries;
+	}
+	
+	public String getAppId() {
+		return appId;
+	}
+	
+	public String getAuthToken() {
+		return authToken;
+	}
+	
+	public static String getBaseUrl() {
+		return BASE_URL;
+	}
+	
+	public CallStatsHttpClient() {
+		super();  
+		
+        // Create I/O reactor configuration
+        IOReactorConfig ioReactorConfig = IOReactorConfig.custom()
+                .setIoThreadCount(Runtime.getRuntime().availableProcessors())
+                .setConnectTimeout(30000)
+                .setSoTimeout(30000)
+                .build();
+
+        // Create a custom I/O reactort
+        ConnectingIOReactor ioReactor;
+		try {
+			ioReactor = new DefaultConnectingIOReactor(ioReactorConfig);
+		} catch (IOReactorException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+			throw new RuntimeException(e);
+		}
+        
+        
+		PoolingHttpClientConnectionManager mgr;
+		mgr = new PoolingHttpClientConnectionManager();
+		((PoolingHttpClientConnectionManager) mgr).setDefaultMaxPerRoute(10);
+		httpclient = HttpClients.custom()
+	            .setConnectionManager(mgr)
+	            .build();
+		
+		PoolingNHttpClientConnectionManager asyncMgr;
+		asyncMgr = new PoolingNHttpClientConnectionManager(ioReactor);
+		httpAsyncClient = HttpAsyncClients.custom()
+				.setConnectionManager(asyncMgr)
+	            .build();
+	}
```

Comment:
I suppose first getHttpClient, then send and receive data. Have you consider  make it a s singleton design pattern, so that customer/developer would not instantiate too many Httpclients. Or do you allow  developers to instantiate many clients?